bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java

```diff
@@ -0,0 +1,185 @@
+    private ConfigurationAdmin admin;
+
+    @Activate
+    public AndroidDebugBridgeDiscoveryService(@Reference ConfigurationAdmin admin) {
+        super(SUPPORTED_THING_TYPES, TIMEOUT_MS, false);
+        this.admin = admin;
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("scan started: searching android devices");
+        discoveryRunning = true;
+        Enumeration<NetworkInterface> nets;
+        Dictionary<String, Object> props = getConfig();
+        if (props == null) {
+            return;
+        }
+        var discoveryPort = Integer.parseInt((String) props.get("discoveryPort"));
+        var discoveryReachableMs = Integer.parseInt((String) props.get("discoveryReachableMs"));
+        var discoveryIpRangeMin = Integer.parseInt((String) props.get("discoveryIpRangeMin"));
+        var discoveryIpRangeMax = Integer.parseInt((String) props.get("discoveryIpRangeMax"));
+        try {
+            nets = NetworkInterface.getNetworkInterfaces();
+            for (NetworkInterface netint : Collections.list(nets)) {
+                Enumeration<InetAddress> inetAddresses = netint.getInetAddresses();
+                for (InetAddress inetAddress : Collections.list(inetAddresses)) {
+                    if (!discoveryRunning)
+                        break;
+                    if (!(inetAddress instanceof Inet4Address))
+                        continue;
+                    if (inetAddress.getHostAddress().equals(LOCAL_INTERFACE_IP))
+                        continue;
+                    String[] ipParts = inetAddress.getHostAddress().split("\\.");
+                    for (var i = discoveryIpRangeMin; i <= discoveryIpRangeMax; i++) {
+                        if (!discoveryRunning)
+                            break;
+                        ipParts[3] = Integer.toString(i);
+                        var currentIp = String.join(".", ipParts);
+                        try {
+                            var currentAddress = InetAddress.getByName(currentIp);
+                            logger.debug("address: {}", currentIp);
+                            if (currentAddress.isReachable(discoveryReachableMs)) {
+                                logger.debug("Reachable ip: {}", currentIp);
+                                var retries = 0;
+                                while (retries < MAX_RETRIES) {
+                                    try {
+                                        discoverWithADB(currentIp, discoveryPort);
+                                    } catch (IOException e) {
+                                        var message = e.getMessage();
+                                        if (message != null && message.contains("rejected by remote peer")) {
+                                            retries++;
+                                            logger.debug("retrying - pending {}", MAX_RETRIES - retries);
+                                            Thread.sleep(5000);
```

Comment:
Please do not perform sleep operations in the `startScan` method. Those are called from a UI thread and as such you should make sure that `startScan` returns quickly.

Reply:
That sleep operation is placed there to allow the user to confirm the host on their device  before a retry and avoid to the user to rerun the entire scan. Either way I can tell that the scan is quite slow because it checks every ip with in the defined range. 
Is it maybe better that I implement the discovery in background? 
I'm not sure how to do it, but maybe you can point me to a binding which uses background discovery.
Let me know your thoughts, if you think that just remove the sleep call is the best option I can go with that.